### PR TITLE
[PHP 8.0] Add constructor promotion

### PR DESCRIPTION
**Pull request by TomasVotruba**
Based on grammar https://github.com/php/php-src/pull/5291/files#diff-7eff82c2c5b45db512a9dc49fb990bb8

- php-src RFC: https://github.com/php/php-src/pull/5291
- wiki RFC: https://wiki.php.net/rfc/constructor_promotion

### CODE REVIEW

**Comment by TomasVotruba** [comment]
Ready for review

**Comment by TomasVotruba** [comment]
I have no idea how to port this part of grammar.

At least, I've added test case that is failing to keep `/** docblock */` above one param:

https://github.com/php/php-src/pull/5291/files#diff-7eff82c2c5b45db512a9dc49fb990bb8R1092

File: lib/PhpParser/PrettyPrinter/Standard.php [hunk 1]
```diff
@@ -725,7 +726,7 @@ protected function pStmt_PropertyProperty(Stmt\PropertyProperty $node) {
     protected function pStmt_ClassMethod(Stmt\ClassMethod $node) {
         return $this->pModifiers($node->flags)
              . 'function ' . ($node->byRef ? '&' : '') . $node->name
-             . '(' . $this->pCommaSeparated($node->params) . ')'
+             . '(' . $this->pMaybeMultiline($node->params) . ')'
```
Comment (TomasVotruba) [inline lib/PhpParser/PrettyPrinter/Standard.php:729]: This seems needed to keep param-per-line formatting.

File: test/code/prettyPrinter/stmt/property_promotion.test [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    }
+}
+-----
+!!php7
+class Point
+{
+    public function __construct(public float $x = 0.0, protected array $y = [], private string $z = 'hello')
+    {
+    }
+}
```
Comment (TomasVotruba) [inline test/code/prettyPrinter/stmt/property_promotion.test:21]: I haven't found a way how to keep this format the same as original one :/

Comment (TomasVotruba) [reply test/code/prettyPrinter/stmt/property_promotion.test:21]: It seems this is never kept in original format: https://github.com/nikic/PHP-Parser/pull/667/commits/e985c914eb0fefacd1499f2cf24aee319a09b36e

So ok :+1:

**Comment by TomasVotruba** [comment]
This is now ready to review & merge

File: lib/PhpParser/Node/Param.php [hunk 3]
```diff
@@ -25,22 +27,26 @@ class Param extends NodeAbstract
...
         $this->byRef = $byRef;
         $this->variadic = $variadic;
         $this->var = $var;
         $this->default = $default;
+        $this->flags = $flags;
     }
 
     public function getSubNodeNames() : array {
-        return ['type', 'byRef', 'variadic', 'var', 'default'];
+        return ['type', 'byRef', 'variadic', 'var', 'default', 'flags'];
```
[the remark this answers is not in the document]
Comment (TomasVotruba) [reply lib/PhpParser/Node/Param.php]: Done

**Comment by nikic** [comment]
Thanks! Merged as https://github.com/nikic/PHP-Parser/commit/b58b19ed1d795d1a3c8056caef295d3045e9d7fd with a change to support formatting-preservation.

**Comment by TomasVotruba** [comment]
Any ETA on releasing this? 
I have a Rector features waiting for this https://github.com/rectorphp/rector/pull/3493

And also PHPStan needs to upgrade to this version before Rector can, as it uses static reflection that prefer it's own classes.

**Comment by nikic** [comment]
@TomasVotruba Done!

**Comment by TomasVotruba** [comment]
Thank you

**Pull request closed without merging**